### PR TITLE
added function to verify a immatriculation number

### PR DESCRIPTION
**Pull request by redalertexpert**
Added functionality to verify immatriculation numbers. Doesn't work until now, always says, that immatriculation numbers are not in the db already. Are students, who are just on waiting lists affected by Registrations.exists()?

Edit: Nvm, works completely

### CODE REVIEW

File: images/spz/spz/templates/internal/registrations.html [hunk 1]
```diff
@@ -22,4 +22,14 @@ <h3 class="ui dividing header">Import</h3>
         <button type="submit" class="fluid ui positive button">Einlesen</button>
     </form>
 </div>
+
+<div class="row">
+    <h3 class="ui dividing header">Matrikelnummer-Überprüfung</h3>
+    <form class="ui form" method="post" action="{{ url_for('registrations_verify') }}" enctype="multipart/form-data">
+        {{ csrf_field() }}
+        {{ render_input(form.number) }}
```
Comment (crepererum) [inline images/spz/spz/templates/internal/registrations.html]: Change that to a string input. The reason for this is that we might want to whitelist specific coworker accounts (for KIT that's "a...") so that they don't have to pay. Also, the tag isn't a number in a traditional sense but rather some kind of ID.

File: images/spz/spz/views.py [hunk 2]
```diff
@@ -190,8 +193,21 @@ def registrations_import():
...
+
+@login_required
+@templated('internal/registrations.html')
+def registrations_verify():
+    form = forms.TagVerifyForm()
+
+    if form.validate_on_submit():
+        num = form.get_tag()
+        if models.verify_tag(num):
+            flash('Matrikelnummer "' + num + '" ist in bereits in der Datenbank', 'success')
```
Comment (crepererum) [inline images/spz/spz/views.py]: Use the template for that, not a flash message.

Comment (redalertexpert) [inline images/spz/spz/views.py]: Which template do you mean? I think everything in this file works with flash messages. If I change that, it would cause inconstistency.

File: images/spz/spz/templates/internal/registrations.html (hunk 1 above)
Comment (redalertexpert) [inline images/spz/spz/templates/internal/registrations.html]: It was a string input, just the naming was confusing.

File: images/spz/spz/forms/__init__.py [hunk 3]
```diff
@@ -511,3 +511,13 @@ class LoginForm(Form):
 
     user = TextField('User', [validators.Required('User muss angegeben werden')])
     password = TextField('Passwort', [validators.Required('Passwort muss angegeben werden')])
+
+class TagVerifyForm(Form):
+    """Represents the form to verify a tag.
+    """
```
Comment (crepererum) [inline images/spz/spz/forms/__init__.py]:
```
"""Represents the form to verify a tag."""
<empty line>
```

File: images/spz/spz/forms/__init__.py [hunk 4]
```diff
@@ -511,3 +511,13 @@ class LoginForm(Form):
...
+    tag = TextField(
+        'Matrikelnummer oder Kürzel'
+    )
+
+    def get_tag(self):
```
Comment (crepererum) [inline images/spz/spz/forms/__init__.py]: not really required (see other forms), but since you already implemented it, it's OK IMHO.

File: images/spz/spz/templates/internal/registrations.html [hunk 5]
```diff
@@ -22,4 +22,14 @@ <h3 class="ui dividing header">Import</h3>
...
     </form>
 </div>
+
+<div class="row">
+    <h3 class="ui dividing header">Matrikelnummer/Kürzel-Überprüfung</h3>
+    <form class="ui form" method="post" action="{{ url_for('registrations_verify') }}" enctype="multipart/form-data">
+        {{ csrf_field() }}
+        {{ render_input(form.tag) }}
+        {{ render_submit(submit='Überprüfen') }}
+
```
Comment (crepererum) [inline images/spz/spz/templates/internal/registrations.html]: remove empty line.

File: images/spz/spz/views.py [hunk 6]
```diff
@@ -145,12 +145,15 @@ def internal():
 @login_required
 @templated('internal/registrations.html')
 def registrations():
-    return None
+
```
Comment (crepererum) [inline images/spz/spz/views.py]: remove empty line.

File: images/spz/spz/templates/internal/registrations.html [hunk 7]
```diff
@@ -22,4 +22,14 @@ <h3 class="ui dividing header">Import</h3>
...
+    </form>
```
Comment (crepererum) [inline images/spz/spz/templates/internal/registrations.html]: Instead of a flash message, the result of the search/request should be appear here.

File: images/spz/spz/forms/__init__.py (hunk 3 above)
Comment (redalertexpert) [inline images/spz/spz/forms/__init__.py]: fixed

File: images/spz/spz/views.py [hunk 8]
```diff
@@ -145,12 +145,14 @@ def internal():
 @login_required
 @templated('internal/registrations.html')
 def registrations():
-    return None
+    form = forms.TagVerifyForm()
+    return dict(form=form)
```
Comment (crepererum) [inline images/spz/spz/views.py]: Does this work w/o `message`? If so, that's fine.

File: images/spz/spz/templates/internal/registrations.html [hunk 9]
```diff
@@ -22,4 +22,16 @@ <h3 class="ui dividing header">Import</h3>
         <button type="submit" class="fluid ui positive button">Einlesen</button>
     </form>
 </div>
+
+<div class="row">
+    <h3 class="ui dividing header">Matrikelnummer/Kürzel-Überprüfung</h3>
+    <form class="ui form" method="post" action="{{ url_for('registrations_verify') }}" enctype="multipart/form-data">
+        {{ csrf_field() }}
+        {{ render_input(form.tag) }}
+        <div class="ui message">
```
Comment (crepererum) [inline images/spz/spz/templates/internal/registrations.html]: I would rather pass `tag_exists` as a template parameter (or smth similar) and generate the message on-the-fly within the template. This way, you could also select the message type (e.g. a green message for "OK" and a red/yellow one for "does not exist).

Another thing to keep the emitted HTML as clean as possible is to avoid the `message` div when no message is passed at all (or when `tag_exists` is `None` when applying the clean design described above). You can use control flow statements within the template, see [the official Jinja2 docs](http://jinja.pocoo.org/docs/dev/templates/#if).

In general, all sentences / texts should be pushed down to the template as good as possible. Flash messages are an exception to avoid over-engineering. Same goes for our log system (otherwise you would have to equip every log message with some kind of enum and more metadata and generate the actual text within the template).

**Comment by crepererum** [comment]
Next iteration. I'm currently rather picky since this is your first contribution and I want to point out some general design principles and guidelines. It doesn't mean that we always follow the rules in that strict way but keeping the code as clean as possible pays off at the end of day (mostly in a way that small feature additions and changes can be realized by small code changes).

File: images/spz/spz/views.py (hunk 8 above)
Comment (redalertexpert) [inline images/spz/spz/views.py]: Here it works.

File: images/spz/spz/templates/internal/registrations.html (hunk 9 above)
Comment (redalertexpert) [inline images/spz/spz/templates/internal/registrations.html]: should be improved now

File: images/spz/spz/templates/internal/registrations.html [hunk 10]
```diff
@@ -22,4 +22,22 @@ <h3 class="ui dividing header">Import</h3>
         <button type="submit" class="fluid ui positive button">Einlesen</button>
     </form>
 </div>
+
+<div class="row">
+    <h3 class="ui dividing header">Matrikelnummer/Kürzel-Überprüfung</h3>
+    <form class="ui form" method="post" action="{{ url_for('registrations_verify') }}" enctype="multipart/form-data">
+        {{ csrf_field() }}
+        {{ render_input(form.tag) }}
+        {% if tag_exists %}
```
Comment (crepererum) [inline images/spz/spz/templates/internal/registrations.html]: As far as I know `None` also evaluates to `False` when being "on its own". So when you pass no object at all [here](https://github.com/spz-signup/spz-signup/pull/62/files#diff-c33e317b7d37b8e87cb84f6fdc929f0aR195), the flow will result in the `else` branch and will try to emit `tag` which ... tada, isn't defined. So it will likely emit "Matrikelnummer/Kürzel "None" ist noch nicht in der Datenbank." as a negative message. This is wrong, since the user didn't request any search at all at this point.

My suggestion is: Replace the `if` condition with a more exact one (`if tag_exists == True`) and the `else` branch with an `elif` and an explicit `elif tag_exisits == False`. This avoids that you need yet another if-guard + indicator variable for the entire construction.

I also might be totally wrong with my analysis of the current commit since I didn't test this, but that's my prediction. It might also just horribly blow up without emitting a misleading message :boom:

**Comment by crepererum** [comment]
One tiny problem left. Rest is perfect :smiley:

File: images/spz/spz/templates/internal/registrations.html [hunk 11]
```diff
@@ -22,4 +22,22 @@ <h3 class="ui dividing header">Import</h3>
...
+        <div class="ui positive message">
+            Matrikelnummer/Kürzel "{{ tag }}" ist in bereits in der Datenbank.
+        </div>
+        {% elif tag.data %}
```
Comment (crepererum) [inline images/spz/spz/templates/internal/registrations.html]: It's very unlikely that this is going to work. If [you don't pass a tag object at all](https://github.com/spz-signup/spz-signup/pull/62/files#diff-c33e317b7d37b8e87cb84f6fdc929f0aR195) then you try to access an attribute of `None` here. Notice you can access objects that are not defined (that results in `None`) but you cannot access attributes of objects that are not defined (= attributes of `None`).

**Comment by crepererum** [comment]
OK, now you tried to fix another bug (which I didn't spot btw) and made things kinda worse. You have to keep in mind 2 edge cases:

1. User presses the submit button w/o entering a tag. This should not result in any message (I missed that part). The lower level API will return `False` in that case since an empty string (or `None`) are both non-existing tags. So you have to check if a tag was entered. You can btw. just issue a redirect to the `/internal/registrations/` route in that case if that makes things easier for you.
2. The user lands on the page via `/internal/registrations/` or `/internal/registrations/import`. In that case, no `tag` and `tag_exists` attributes are passed at all to the template (at least in your current implementation). No message should appear. Be aware that both objects are `None` and you cannot use any attributes / methods.

File: images/spz/spz/templates/internal/registrations.html (hunk 11 above)
Comment (redalertexpert) [inline images/spz/spz/templates/internal/registrations.html]: tag.data is bs. It should be simply "tag"

**Comment by crepererum** [comment]
OK, that is functionally correct now. IMHO the template code doesn't look very clean though. The control flow is:

```python
if tag_exists:
  # case a)
  # ok, some tag exists, so we emit a green message
elif tag:
  # case b)
  # wait, what is going on here? from the first `if`,
  # we know that the tag does not exist but for some reason we check
  # if `tag` is not `None`/`False`/`0`/...
  # What exactly is the implication of a readable`tag` variable with a tag that does not exist?!
  # Why is there no case c)?
```

Something more understandable would be:

```python
if tag:
  # OK, so there is a readable tag variable and this block is going to
  # argue about the tag details
  if tag_exists:
    # green message
  else:
    # red message
# no case c), but more obvious why
```

**Comment by crepererum** [comment]
That looks perfect.

A note about handling of commits in PRs: I really like that you add new commits whenever you do another iteration. That helps reviewing changes A LOT (otherwise it would be really hard to reason about "changes of changes" w/o using yet another tool like https://reviewable.io/). Sadly, this has the drawback that the commits structure does not symbolize the real change anymore. As you might have guessed, there is a good solution for that too: AFTER the PR was approved (= someone saying the changes are finally OK, NOT merging / committing them to `master`), you do a final iteration that rebases the entire branch and squashes the commits (in one single or multiple atomic commits, depending on the PR [1 single in this case]). For other projects there would be a final review for that one but because you're a main contributor here and I trust you, you can just push the resulting changes afterwards w/o yet-another-OK from myself.

Keep in mind to squash / amend / change commits before pushing them to `master`, since for this projects this is a [protected branch](https://help.github.com/articles/about-protected-branches/) (so you're not allowed to do forced pushes and other funny things that changes/deletes commits that where already commited). That protection is only activated for `master` since it doesn't make sense for PRs. Obviously this protection is only enforced on GitHub. Locally you can still mess around with protected branches.

TL;DR; rebase + squash + push (+ close) :+1:

**Comment by redalertexpert** [comment]
rebased and pushed
